src/plugins/kernel_install/utils.sh

```diff
@@ -550,32 +550,85 @@ function do_uninstall()
   fi
 }
 
+# Element existence check
+# Checks if element belongs to given array
+#
+# @target_element element being checked
+# @array target array
+# returns 1 if element is in array and 0 otherwise
+function is_in_array()
+{
+  local target_element="$1"
+  local array=("${!2}")
+
+  for element in "${array[@]}"; do
+    if [[ "$element" == "$target_element" ]]; then
+      return 1
+    fi
+  done
+  return 0
+}
+
+# Return available kernels
+#
+# @prefix Set a base prefix for searching for kernels.
+# @all_kernels List all available kernels, not only the ones installed by kw
+# returns array containing available kernels
+function process_installed_kernels()
+{
+  local all_kernels="$1"
+  local prefix="$2"
+  local kernels
+
+  kernels=$(list_installed_kernels 'SILENT' 1 "$all_kernels" "$prefix")
+  IFS=, read -r -a available_kernels <<< "$kernels"
+  mapfile -t deduped_kernels < <(printf "%s\n" "${available_kernels[@]}" | sort --unique)
+}
+
 function kernel_uninstall()
 {
   local reboot="$1"
   local target="$2"
-  local kernel_list_string="$3"
+  local kernel_list_string_or_regex="$3"
   local flag="$4"
   local force="$5"
   local prefix="$6"
   local update_grub=0
+  declare -A kernel_names
+
+  prefix_for_regex="regex:"
 
   cmd_manager "$flag" "sudo mkdir -p $REMOTE_KW_DEPLOY"
   cmd_manager "$flag" "sudo touch '$INSTALLED_KERNELS_PATH'"
 
-  kernel_list_string=$(printf '%s' "$kernel_list_string" | tr --delete ' ')
+  process_installed_kernels 1 "$prefix"
+  deduped_all_kernels=("${deduped_kernels[@]}")
+
+  process_installed_kernels '' "$prefix"
+  deduped_kw_managed_kernels=("${deduped_kernels[@]}")
 
-  if [[ -z "$kernel_list_string" ]]; then
+  if [[ -z "$kernel_list_string_or_regex" ]]; then
     printf '%s\n' 'Invalid argument'
     exit 22 #EINVAL
   fi
```

Comment:
Move this check to before the processing, as it will be meaningless if it fails anyway.